src/test/java/marquez/db/DatasetDaoTest.java

```diff
@@ -77,13 +100,36 @@ public void testInsert() {
     assertThat(rowsAfter).isEqualTo(rowsBefore + 1);
   }
 
+  @Test(expected = UnableToExecuteStatementException.class)
+  public void testIdempotentInsert() {
+    final int rowsBefore = datasetDao.count();
+
+    final DatasetRow newDatasetRow =
+        newDatasetRowWith(namespaceRow.getUuid(), datasourceRow.getUuid());
+    datasetDao.insert(newDatasetRow);
+
+    final int rowsAfter = datasetDao.count();
+    assertThat(rowsAfter).isEqualTo(rowsBefore + 1);
+
+    datasetDao.insert(newDatasetRow);
+  }
+
+  @Test(expected = UnableToExecuteStatementException.class)
+  public void testIdempotentInsertAndGet() {
+    final int rowsBefore = datasetDao.count();
+
+    final DatasetRow newDatasetRow =
+        newDatasetRowWith(namespaceRow.getUuid(), datasourceRow.getUuid());
+    datasetDao.insert(newDatasetRow);
```

Comment:
You're testing `DatasetDao.insert()`. Is this test even necessary for `DatasetDao.insertAndGet()`?

Reply:
This is part of the setup and not what's under test. Since we're not validating the data that's inserted, there's no purpose to get the data. Seems confusing to `get` anything and then not store it.

That said, this test is necessary because `insert` and `insertAndGet` follow different code paths and issue different queries. It was that divergent behavior (one had `ON CONFLICT DO NOTHING` and one did not) that led me to write this test case in the first place. While it's true that similar now, but if they diverged in an erroneous way, having a test for each would help detect that, where just one test could miss it.

Reply:
Make sense 👍